x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java

```diff
@@ -116,6 +131,21 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeBoolean(synthetic());
     }
 
+    public static FieldAttribute readFrom(StreamInput in) throws IOException {
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_FIELD_ATTRIBUTE_CACHED_SERIALIZATION)
+            && in instanceof PlanStreamInput psi) {
+            boolean inCache = in.readBoolean();
+            int cacheId = in.readInt();
+            if (inCache) {
+                return psi.attributeFromCache(cacheId);
```

Comment:
Question: I guess we can assume that during deserialization, we walk the tree in the same order when serializing? Otherwise we might find an attribute here that's already _supposed_ to be in our cache, but we didn't encounter the explicit version yet?

Reply:
Theoretically, we _could_ have problems with circular dependencies (loops in the parent/child relationship), but we would have the same problem also with the previous serialization.
Actually, we'd throw a StackOveflowError in both cases.

Practically, I don't think it will ever happen.

Reply:
I'm not super worried about an unintentional loop here. If we were shipping this code to space, yeah, no good. But we don't build loops of attributes so I think we're ok.

Reply:
What I meant is not about loops, but about deserializing and encountering an id for a cached attribute before we have encountered the actual attribute. I _don't think_ this can happen at all, because we don't do random seeks in the stream during deserialization (right?), so we cannot read the cache id of an attribute before we encounter the actual attribute belonging to this cache id.

Loops with the parent-child relationship in fieldattributes will blow up during serialization even in the current state, I think, because that'll lead to an infinite loop.